### PR TITLE
Remove deprecated call to DialContext in Hubble server

### PR DESCRIPTION
**Pull request by davchos**
Remove deprecated call to DialContext in Hubble server

Related:#32274, #33065

Signed-off-by: David Chosrova <dchosrova@gmail.com>

### CODE REVIEW

**Comment by davchos** [comment]
Hi @tklauser,

I have 2 questions, before pushing this PR, I have used make kind / make kind-image-fast / make kind-install-cilium-fast
to test if everything was OK. I don't understand why the import not used has not been catched. 

I have sent a commit to remove  the unused  import, the check is still failing, so is it best now to close this PR and re submit a new one.

Thanks in advance for your help.

David

File: pkg/hubble/peer/types/client.go
Comment (glrf) [inline pkg/hubble/peer/types/client.go:53]: As far as I can tell the `DialTimeout` is not completely unused, so IMO we can remove it.

File: pkg/hubble/peer/types/client.go
Comment (glrf) [inline pkg/hubble/peer/types/client.go:95]: Same here for the `RemoteClientBuilder`. The DialTimeout is now unused and can be removed

File: pkg/hubble/relay/pool/client.go
Comment (glrf) [inline pkg/hubble/relay/pool/client.go:50]: And again here. We don't use the `DialTimeout` anymore, so let's remove it from the `GRPCClientConnBuilder` too.

File: pkg/hubble/relay/server/server.go
Comment (glrf) [inline pkg/hubble/relay/server/server.go]: With all the unused `DialTimeout`s I think we can then actually get rid of the `dialTimeout` option. Which would then even mean that we can get rid of the the `--dial-timeout` flag for relay

**Review by glrf — COMMENTED** [review]
Hi David thanks for the contribution!

For you're two questions:

1. The reason why `make kind-image-fast` didn't catch the issue is because it doesn't actually build the relay image. So it wasn't actually compiling your code changes. Sadly, the Hubble Relay development isn't quite a streamlined, it's not even enabled by default when you use `make kind / make kind-image-fast / make kind-install-cilium-fast`.

   To actually run your Relay changes you need to build the image yourself and enable relay by running the following after you created the kind cluster

```
DOCKER_IMAGE_TAG=local make docker-hubble-relay-image
kind load docker-image quay.io/cilium/hubble-relay:local
cat << EOF > contrib/testing/kind-custom.yaml
hubble:
  relay:
    enabled: true
    image:
      override: quay.io/cilium/hubble-relay:local
      pullPolicy: Never
EOF
make kind-install-cilium-fast
```


2. The reason CI is still failing is that we actually require that every commit builds successfully, not just the last commit. You can just squash your 4 commits into a single commit and CI should be happy again.


On a first look your changes look good, but I think we can actually get rid of the `DialTimeout` and the related option and flag as it doesn't do anything anymore (and then even remove it from the helm chart)

I still want take a closer look how this will impact the peer manager though, just to be sure we don't break anything unexpectedly by removing this timeout. But I won't get to that today, so I'll give you another review on Monday.

**Comment by glrf** [comment]
Alright, I took another look and I think removing this timeout is fine. It won't change anything for the connection to the peers and connection retries to the peer service will have a slightly different timing, because we don't have the same timeout when dialing on the first call, but that should be fine.

So let's just remove the unused `DialTimeout` fields and options and the corresponding flag and this should be good.

**Comment by davchos** [comment]
> Alright, I took another look and I think removing this timeout is fine. It won't change anything for the connection to the peers and connection retries to the peer service will have a slightly different timing, because we don't have the same timeout when dialing on the first call, but that should be fine.
> 
> So let's just remove the unused `DialTimeout` fields and options and the corresponding flag and this should be good.

Hi Fabian,

I will  remove the timeout, not available today I do it tomorrow. Have a good day.

David

**Comment by davchos** [comment]
No idea why this test is failing, the cilium installation look ok, Hubble relay not enabled. 🥲

File: install/kubernetes/cilium/templates/hubble-relay/configmap.yaml
Comment (glrf) [inline install/kubernetes/cilium/templates/hubble-relay/configmap.yaml:31]: Could you add a note to the helm value that it is deprecated and a no-op.

**Review by glrf — COMMENTED** [review]
One final nitpick, otherwise this LGTM

And the CI failure seems like an unrelated flake.

**Review by glrf — APPROVED** [review]
LGTM :rocket:

**Comment by glrf** [comment]
Sadly, that one does seem related: https://github.com/cilium/cilium/actions/runs/10382885529/job/28746842560

We introduce a small issue during startup. When the DNS server is not (yet) available, `DialContext` blocked until the connections was ready or the timeout was reached. Now we do the resolving during the `Notify` call, which will instantly fail, assuming we'll retry ourselves. Which we'll do, but we'll log a warning. This means on a clean Cilium installation Hubble Relay will always log a warning. Which isn't horrible, but also not great and what currently fails CI.

IMO for now it would be fine to just drop the severity of [that log message](https://github.com/cilium/cilium/blob/main/pkg/hubble/relay/pool/manager.go#L117) from `Warning` to `Info`. We're actively retrying and it will eventually succeed and is more or less what happened before, we just didn't log anything for 30 seconds. 

Longer term we might want to take a closer look at the Peer service connection and how we handle retries and make something smarter than just tearing down everything and retrying in 30 seconds. But I think that's out of scope of this PR and that's something I wanted to do for a while now anyways.

**Comment by davchos** [comment]
> Sadly, that one does seem related: https://github.com/cilium/cilium/actions/runs/10382885529/job/28746842560
> 
> We introduce a small issue during startup. When the DNS server is not (yet) available, `DialContext` blocked until the connections was ready or the timeout was reached. Now we do the resolving during the `Notify` call, which will instantly fail, assuming we'll retry ourselves. Which we'll do, but we'll log a warning. This means on a clean Cilium installation Hubble Relay will always log a warning. Which isn't horrible, but also not great and what currently fails CI.
> 
> IMO for now it would be fine to just drop the severity of [that log message](https://github.com/cilium/cilium/blob/main/pkg/hubble/relay/pool/manager.go#L117) from `Warning` to `Info`. We're actively retrying and it will eventually succeed and is more or less what happened before, we just didn't log anything for 30 seconds.
> 
> Longer term we might want to take a closer look at the Peer service connection and how we handle retries and make something smarter than just tearing down everything and retrying in 30 seconds. But I think that's out of scope of this PR and that's something I wanted to do for a while now anyways.

Thanks for the explanation,  I'll change the log level.

File: install/kubernetes/cilium/values.yaml.tmpl
Comment (squeed) [inline install/kubernetes/cilium/values.yaml.tmpl:1562]: one more thing: can you add a note in `Documentation/operations/upgrade.rst` please? Thanks!

Comment (davchos) [reply install/kubernetes/cilium/values.yaml.tmpl:1562]: Yes, I can. I'll do it on Monday. Have a nice weekend.

File: Documentation/operations/upgrade.rst
Comment (learnitall) [inline Documentation/operations/upgrade.rst]:
```suggestion
* The ``hubble.relay.dialTimeout`` helm option has been deprecated and is now a no-op
```

**Review by learnitall — CHANGES_REQUESTED** [review]
Thanks! One note for the upgrade notes to make it more specific, but otherwise LGTM.

**Comment by davchos** [comment]
Hi,

After running locally the "make -C bpf checkpatch" I get errors, "Warning: WARNING:BAD_SIGN_OFF: Duplicate signature".

Maybe it is better to  close this PR and re create a clean one ? 

David

File: Documentation/operations/upgrade.rst
Comment (learnitall) [inline Documentation/operations/upgrade.rst]:
```suggestion
* The ``hubble.relay.dialTimeout`` helm option has been deprecated and is now a no-op.
```

**Review by learnitall — APPROVED** [review]
I found a non-blocking nit, but LGTM. Thank you!

**Comment by glrf** [comment]
> Maybe it is better to close this PR and re create a clean one ?

That shouldn't be necessary. I think the issue is related to the merge commits from merging main into this branch. At last merge commit isn't signed off and I think our tooling in general assumes that we rebase instead of merge to update branches.

Could you please drop your two merge commits and then rebase on main? That should fix the issue.

**Comment by davchos** [comment]
@glrf , @squeed, @learnitall @kaworu,
Thanks for walking me through this commit, I really appreciate

File: pkg/hubble/peer/types/client.go
Comment (squeed) [inline pkg/hubble/peer/types/client.go:95]: You can recreate this timeout by doing:
```
ctx, cancel := context.WithTimeout(...)
defer cancel()
conn.WaitForStateChange(ctx, connectivity.Ready)
```

and keep the dial timeout (as well as preventing spurious error messages). Seem reasonable?

Comment (glrf) [reply pkg/hubble/peer/types/client.go:95]: Sorry, this PR fell off my radar.

I agree the log change is a hack and probably not the correct approach, I just didn't want block this PR because of random warning logs.

But also IMO there isn't really a point in keeping this timeout, I doubt anyone ever configured this, as it just slightly changes the timing of reconnecting. And blocking dials are explicitly an anti-pattern (https://github.com/grpc/grpc-go/blob/master/Documentation/anti-patterns.md#especially-bad-using-deprecated-dialoptions), so don't really want to "re-implement" them.

Also `WaitForStateChange` is still marked as experimental, so I'm not sure it's significantly better to use that instead of the deprecated calls. And BTW it's not quite as simple because `WaitForStateChange` will only tell you when the state changed *from* the state you provide, no *to* the state..

IMO the PeerClient setup would need some refactoring to properly solve this where we:

1. Don't close the connection on any error. There is no point in doing that, gRPC will redial the connection if necessary. So only create the client once and reuse it
2. Do something smarter than just sleeping for 30 seconds on any failure and retrying. We should either use an exponential backoff, like we do for ClientConns, or see if we can introduce proper retries (https://github.com/grpc/grpc-go/blob/master/examples/features/retry/README.md)
3. Potentially implement some initial delay where we don't log errors. But this probably wouldn't be necessary when using proper gRPC retries.

But this is IMO out-of-scope for this PR, so I didn't want to propose it at first, but I might do it as a follow up myself.

Comment (davchos) [reply pkg/hubble/peer/types/client.go:95]: Hi,

For the third point, I have seen in the spire code that they check if the service they rely on is resolvable. Might be  better than a delay ? 

https://github.com/cilium/cilium/blob/fc86dfef89ae06478468fde7ca39786dd569b76d/operator/auth/spire/client.go#L317

I'll try this week to test it.

David

**Review by kaworu — APPROVED** [review]

**Comment by kaworu** [comment]
Patch LGTM, thanks @glrf for following up and reviewing. Added the `dont-merge/blocked` label as IIUC the author still wishes to investigate https://github.com/cilium/cilium/pull/34261#discussion_r1789695437.

File: pkg/hubble/peer/types/client.go
Comment (dylandreimerink) [reply pkg/hubble/peer/types/client.go:95]: Any progress on this?

**Comment by derailed** [comment]
@davchos Could you address the conflicts when you get a chance? Thank you!

**Comment by davchos** [comment]
Well I think it is best to close this PR and work on a new one. If I find time I will try to resubmit a new one.

**Pull request closed without merging**
